File: agenkit-ts/examples/src/techniques/reasoning/self-consistency.js
"use strict";
/**
 * Self-Consistency Reasoning Technique
 *
 * Self-Consistency improves reliability by generating multiple independent reasoning
 * paths and using voting to select the most consistent answer.
 *
 * Reference: "Self-Consistency Improves Chain of Thought Reasoning in Language Models"
 * Wang et al., 2022 - https://arxiv.org/abs/2203.11171
 *
 * Key features:
 * - Parallel sampling for performance
 * - Multiple voting strategies (majority, weighted, first)
 * - Customizable answer extraction
 * - Consistency scoring for confidence measurement
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelfConsistencyAgent = void 0;
exports.createSelfConsistencyAgent = createSelfConsistencyAgent;
const interfaces_1 = require("../../core/interfaces");
/**
 * Default answer extractor that looks for common answer patterns.
 *
 * Patterns recognized:
 * - "Therefore, X" / "Thus, X" / "So, X"
 * - "The answer is X"
 * - "= X" (for math)
 * - "Conclusion: X" / "Result: X"
 * - Last non-empty line (fallback)
 */
function defaultAnswerExtractor(text) {
    // Try explicit answer markers
    const patterns = [
        /(?:therefore|thus|so),?\s+(?:the answer is\s+)?(.+?)(?:\.|$)/i,
        /(?:the answer is|answer:)\s+(.+?)(?:\.|$)/i,
        /=\s*(.+?)(?:\n|$)/,
        /(?:conclusion|result):\s*(.+?)(?:\.|$)/i,
    ];
    for (const pattern of patterns) {
        const match = text.match(pattern);
        if (match) {
            return match[1].trim();
        }
    }
    // Fallback: use last non-empty line
    const lines = text.split('\n');
    for (let i = lines.length - 1; i >= 0; i--) {
        const line = lines[i].trim();
        if (line) {
            return line;
        }
    }
    return text.trim();
}
/**
 * Self-Consistency agent that wraps a base agent.
 *
 * Usage:
 * ```typescript
 * const sc = new SelfConsistencyAgent(baseAgent, {
 *   numSamples: 5,
 *   votingStrategy: 'majority',
 * });
 *
 * const response = await sc.process(message);
 * console.log(`Consensus: ${response.content}`);
 * console.log(`Confidence: ${response.metadata.consistency_score}`);
 * ```
 */
class SelfConsistencyAgent {
    constructor(agent, config = {}) {
        this.agent = agent;
        this.name = 'self_consistency';
        this.capabilities = [
            'reasoning',
            'self_consistency',
            'majority_voting',
            'reliability',
            'consensus',
        ];
        this.numSamples = config.numSamples ?? 5;
        this.votingStrategy = config.votingStrategy ?? 'majority';
        this.temperature = config.temperature;
        this.answerExtractor = config.answerExtractor ?? defaultAnswerExtractor;
    }
    /**
     * Process a message with Self-Consistency.
     *
     * Generates multiple independent samples, extracts answers, and uses
     * voting to determine the most consistent answer.
     */
    async process(message) {
        // Generate multiple samples in parallel
        const { fullResponses, extractedAnswers } = await this.generateSamples(message);
        // Vote for consensus answer
        let consensusAnswer;
        let consistencyScore;
        switch (this.votingStrategy) {
            case 'majority':
                ({ answer: consensusAnswer, score: consistencyScore } = this.voteMajority(extractedAnswers));
                break;
            case 'weighted':
                ({ answer: consensusAnswer, score: consistencyScore } = this.voteWeighted(extractedAnswers, fullResponses));
                break;
            case 'first':
                ({ answer: consensusAnswer, score: consistencyScore } = this.voteFirst(extractedAnswers));
                break;
            default:
                throw new Error(`Invalid voting strategy: ${this.votingStrategy}`);
        }
        // Count answer occurrences for metadata
        const answerCounts = this.countAnswers(extractedAnswers);
        // Build response with metadata
        return (0, interfaces_1.createMessage)('assistant', consensusAnswer, {
            technique: 'self_consistency',
            num_samples: this.numSamples,
            voting_strategy: this.votingStrategy,
            consistency_score: consistencyScore,
            samples: fullResponses,
            extracted_answers: extractedAnswers,
            answer_counts: answerCounts,
            base_agent: this.agent.name,
        });
    }
    /**
     * Generate multiple samples in parallel.
     */
    async generateSamples(message) {
        // Generate samples in parallel
        const samplePromises = Array.from({ length: this.numSamples }, () => this.sampleOnce(message));
        try {
            const samples = await Promise.all(samplePromises);
            const fullResponses = samples.map(s => s.fullResponse);
            const extractedAnswers = samples.map(s => s.extractedAnswer);
            return { fullResponses, extractedAnswers };
        }
        catch (error) {
            throw new Error(`Sampling failed: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    /**
     * Generate one sample from the base agent.
     */
    async sampleOnce(message) {
        // TODO: If temperature supported, pass it to agent
        const response = await this.agent.process(message);
        const fullResponse = String(response.content);
        const extractedAnswer = this.answerExtractor(fullResponse);
        return { fullResponse, extractedAnswer };
    }
    /**
     * Vote using majority (most common answer wins).
     */
    voteMajority(answers) {
        if (answers.length === 0) {
            return { answer: '', score: 0.0 };
        }
        // Count answer occurrences (case-insensitive)
        const counts = new Map();
        const originalCase = new Map();
        for (const answer of answers) {
            const normalized = answer.toLowerCase().trim();
            counts.set(normalized, (counts.get(normalized) || 0) + 1);
            if (!originalCase.has(normalized)) {
                originalCase.set(normalized, answer);
            }
        }
        // Find most common
        let winningAnswer = '';
        let maxCount = 0;
        for (const [normalized, count] of Array.from(counts.entries())) {
            if (count > maxCount) {
                maxCount = count;
                winningAnswer = normalized;
            }
        }
        // Get original case version
        const winner = originalCase.get(winningAnswer) || winningAnswer;
        const consistencyScore = maxCount / answers.length;
        return { answer: winner, score: consistencyScore };
    }
    /**
     * Vote using weighted strategy (longer responses get more weight).
     */
    voteWeighted(answers, responses) {
        if (answers.length === 0) {
            return { answer: '', score: 0.0 };
        }
        // Group answers by normalized form
        const groups = new Map();
        for (let i = 0; i < answers.length; i++) {
            const normalized = answers[i].toLowerCase().trim();
            const existing = groups.get(normalized);
            if (existing) {
                existing.weight += responses[i].length;
                existing.count += 1;
            }
            else {
                groups.set(normalized, {
                    original: answers[i],
                    weight: responses[i].length,
                    count: 1,
                });
            }
        }
        // Find highest weighted answer
        let winningAnswer = '';
        let maxWeight = 0;
        let totalWeight = 0;
        for (const group of Array.from(groups.values())) {
            totalWeight += group.weight;
            if (group.weight > maxWeight) {
                maxWeight = group.weight;
                winningAnswer = group.original;
            }
        }
        const consistencyScore = totalWeight > 0 ? maxWeight / totalWeight : 0.0;
        return { answer: winningAnswer, score: consistencyScore };
    }
    /**
     * Use first answer (no voting, for debugging).
     */
    voteFirst(answers) {
        if (answers.length === 0) {
            return { answer: '', score: 0.0 };
        }
        return { answer: answers[0], score: 1.0 };
    }
    /**
     * Count answer occurrences (case-insensitive).
     */
    countAnswers(answers) {
        const counts = {};
        for (const answer of answers) {
            const normalized = answer.toLowerCase().trim();
            counts[normalized] = (counts[normalized] || 0) + 1;
        }
        return counts;
    }
}
exports.SelfConsistencyAgent = SelfConsistencyAgent;
/**
 * Factory function to create a Self-Consistency agent.
 *
 * @param agent Base agent to wrap
 * @param config Configuration options
 * @returns Self-Consistency agent
 */
function createSelfConsistencyAgent(agent, config) {
    return new SelfConsistencyAgent(agent, config);
}

File: agenkit/techniques/reasoning/self_consistency.py
"""
Self-Consistency Reasoning Technique

Generates multiple independent reasoning paths and selects the most consistent
answer through voting or aggregation strategies.

This technique improves reliability by sampling multiple times and using
consensus, particularly effective when multiple valid reasoning approaches exist.

References:
    - Paper: https://arxiv.org/abs/2203.11171 (Wang et al., 2022)
    - "Self-Consistency Improves Chain of Thought Reasoning in Language Models"
    - Works best combined with CoT or other reasoning techniques

Example:
    Basic usage::

        from agenkit.techniques.reasoning import ChainOfThought, SelfConsistency
        from agenkit import Message

        # Wrap CoT with Self-Consistency
        cot = ChainOfThought(llm=my_llm)
        sc = SelfConsistency(
            agent=cot,
            num_samples=5,
            voting_strategy="majority"
        )

        response = await sc.process(Message(
            role="user",
            content="What is 15 * 24?"
        ))

        # Access consensus info
        print(f"Consensus answer: {response.content}")
        print(f"Consistency score: {response.metadata['consistency_score']}")
"""

import asyncio
import re
from collections import Counter
from collections.abc import Callable

from agenkit import Agent, Message


class SelfConsistency(Agent):
    """
    Self-Consistency reasoning technique.

    Wraps a base agent, samples multiple times, and uses voting to determine
    the most consistent answer. Improves reliability by leveraging consensus
    across diverse reasoning paths.

    This technique is particularly effective for:
    - Tasks with multiple valid solution approaches
    - Improving answer reliability and confidence
    - Reducing impact of individual reasoning errors
    - Tasks where consistency indicates correctness

    Attributes:
        name: Agent name (always "self_consistency")
        agent: Base agent to sample from
        num_samples: Number of independent samples to generate
        voting_strategy: Strategy for aggregating answers
        temperature: Sampling temperature for diversity
        answer_extractor: Function to extract final answer from response
    """

    def __init__(
        self,
        agent: Agent,
        num_samples: int = 5,
        voting_strategy: str = "majority",
        temperature: float | None = None,
        answer_extractor: Callable[[str], str] | None = None,
    ):
        """
        Initialize Self-Consistency agent.

        Args:
            agent: Base agent to sample from. Can be any Agent (CoT, ToT, etc.)
            num_samples: Number of independent reasoning paths to generate.
                Higher values improve reliability but increase cost. Default is 5.
            voting_strategy: How to aggregate multiple answers:
                - "majority": Select most common answer (default)
                - "weighted": Weight by answer confidence/length
                - "first": Use first answer (no voting, for debugging)
            temperature: Optional temperature for sampling diversity. If provided,
                passed to agent if it supports temperature parameter.
            answer_extractor: Custom function to extract final answer from
                response text (str -> str). If None, uses default extraction
                that looks for common answer patterns.

        Example:
            >>> from agenkit.techniques.reasoning import ChainOfThought
            >>> cot = ChainOfThought(llm=my_llm)
            >>> sc = SelfConsistency(
            ...     agent=cot,
            ...     num_samples=7,
            ...     voting_strategy="majority"
            ... )
        """
        self.agent = agent
        self.num_samples = num_samples
        self.voting_strategy = voting_strategy
        self.temperature = temperature
        self.answer_extractor = answer_extractor or self._default_answer_extractor

    @property
    def name(self) -> str:
        """Return agent name."""
        return "self_consistency"

    def _default_answer_extractor(self, text: str) -> str:
        """
        Extract final answer from response text.

        Looks for common answer patterns:
        - "Therefore, X" / "Thus, X" / "So, X"
        - "The answer is X"
        - "= X" (for math)
        - Last sentence/line (fallback)

        Args:
            text: Response text to extract answer from

        Returns:
            Extracted answer string
        """
        # Try explicit answer markers
        answer_patterns = [
            r"(?:therefore|thus|so),?\s+(?:the answer is\s+)?(.+?)(?:\.|$)",
            r"(?:the answer is|answer:)\s+(.+?)(?:\.|$)",
            r"=\s*(.+?)(?:\n|$)",
            r"(?:conclusion|result):\s*(.+?)(?:\.|$)",
        ]

        for pattern in answer_patterns:
            match = re.search(pattern, text, re.IGNORECASE | re.MULTILINE)
            if match:
                return match.group(1).strip()

        # Fallback: use last non-empty line
        lines = [line.strip() for line in text.split("\n") if line.strip()]
        if lines:
            return lines[-1]

        return text.strip()

    async def _sample_once(self, message: Message) -> tuple[str, str]:
        """
        Generate one sample from base agent.

        Args:
            message: Input message

        Returns:
            Tuple of (full_response, extracted_answer)
        """
        # TODO: If temperature supported, pass it to agent
        response = await self.agent.process(message)
        full_response = response.content
        answer = self.answer_extractor(full_response)
        return (full_response, answer)

    async def _generate_samples(self, message: Message) -> list[tuple[str, str]]:
        """
        Generate multiple samples in parallel.

        Args:
            message: Input message

        Returns:
            List of (full_response, extracted_answer) tuples
        """
        # Generate samples in parallel for performance
        tasks = [self._sample_once(message) for _ in range(self.num_samples)]
        samples = await asyncio.gather(*tasks)
        return samples

    def _vote_majority(self, answers: list[str]) -> tuple[str, float]:
        """
        Majority voting: select most common answer.

        Args:
            answers: List of extracted answers

        Returns:
            Tuple of (winning_answer, consistency_score)
        """
        if not answers:
            return ("", 0.0)

        # Count answer occurrences (case-insensitive)
        answer_counts = Counter(answer.lower().strip() for answer in answers)

        # Get most common
        most_common = answer_counts.most_common(1)[0]
        winning_answer_lower = most_common[0]
        winning_count = most_common[1]

        # Find original case version
        winning_answer = next(
            (ans for ans in answers if ans.lower().strip() == winning_answer_lower),
            winning_answer_lower,
        )

        # Consistency score = fraction that agree
        consistency_score = winning_count / len(answers)

        return (winning_answer, consistency_score)

    def _vote_weighted(self, answers: list[str], responses: list[str]) -> tuple[str, float]:
        """
        Weighted voting: weight by response length (proxy for detail/confidence).

        Args:
            answers: List of extracted answers
            responses: List of full response texts

        Returns:
            Tuple of (winning_answer, consistency_score)
        """
        if not answers:
            return ("", 0.0)

        # Group answers by normalized form
        answer_groups = {}
        for answer, response in zip(answers, responses, strict=False):
            answer_key = answer.lower().strip()
            if answer_key not in answer_groups:
                answer_groups[answer_key] = {"original": answer, "weight": 0, "count": 0}
            # Weight by response length
            answer_groups[answer_key]["weight"] += len(response)
            answer_groups[answer_key]["count"] += 1

        # Find highest weighted answer
        winning_key = max(answer_groups, key=lambda k: answer_groups[k]["weight"])
        winning_answer = answer_groups[winning_key]["original"]
        winning_weight = answer_groups[winning_key]["weight"]
        total_weight = sum(group["weight"] for group in answer_groups.values())

        # Consistency score = weight fraction
        consistency_score = winning_weight / total_weight if total_weight > 0 else 0.0

        return (winning_answer, consistency_score)

    def _vote_first(self, answers: list[str]) -> tuple[str, float]:
        """
        First answer (no voting): use first sample.

        Args:
            answers: List of extracted answers

        Returns:
            Tuple of (first_answer, 1.0)
        """
        if not answers:
            return ("", 0.0)
        return (answers[0], 1.0)

    async def process(self, message: Message) -> Message:
        """
        Process message with Self-Consistency.

        Generates multiple independent samples, extracts answers, and uses
        voting to determine the most consistent answer.

        Args:
            message: Input message with query content

        Returns:
            Message with consensus answer and metadata. Metadata includes:
                - samples: List of all full responses
                - extracted_answers: List of extracted answers
                - consistency_score: Agreement score (0.0-1.0)
                - num_samples: Number of samples generated
                - voting_strategy: Strategy used
                - answer_counts: Dict of answer frequencies
                - technique: Always "self_consistency"

        Example:
            >>> response = await sc.process(Message(role="user", content="What is 2+2?"))
            >>> print(f"Consensus: {response.content}")
            >>> print(f"Consistency: {response.metadata['consistency_score']:.2f}")
        """
        # Generate multiple samples
        samples = await self._generate_samples(message)
        full_responses = [resp for resp, _ in samples]
        extracted_answers = [ans for _, ans in samples]

        # Vote for consensus answer
        if self.voting_strategy == "majority":
            consensus_answer, consistency_score = self._vote_majority(extracted_answers)
        elif self.voting_strategy == "weighted":
            consensus_answer, consistency_score = self._vote_weighted(
                extracted_answers, full_responses
            )
        elif self.voting_strategy == "first":
            consensus_answer, consistency_score = self._vote_first(extracted_answers)
        else:
            raise ValueError(f"Invalid voting strategy: {self.voting_strategy}")

        # Count answer occurrences for metadata
        answer_counts = Counter(ans.lower().strip() for ans in extracted_answers)

        return Message(
            role="assistant",
            content=consensus_answer,
            metadata={
                "technique": "self_consistency",
                "num_samples": self.num_samples,
                "voting_strategy": self.voting_strategy,
                "consistency_score": consistency_score,
                "samples": full_responses,
                "extracted_answers": extracted_answers,
                "answer_counts": dict(answer_counts),
                "base_agent": self.agent.name if hasattr(self.agent, "name") else "unknown",
            },
        )

    @property
    def capabilities(self) -> list[str]:
        """
        Return agent capabilities.

        Returns:
            List of capability strings describing what this agent can do
        """
        return ["reasoning", "self_consistency", "majority_voting", "reliability", "consensus"]
